server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+import org.apache.logging.log4j.util.Supplier;
+import org.opensearch.action.admin.indices.tiering.HotToWarmTieringResponse;
+import org.opensearch.action.admin.indices.tiering.TieringIndexRequest;
+import org.opensearch.action.admin.indices.tiering.TieringRequestContext;
+import org.opensearch.cluster.ClusterChangedEvent;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.ClusterStateListener;
+import org.opensearch.cluster.ClusterStateUpdateTask;
+import org.opensearch.cluster.NotClusterManagerException;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.routing.RoutingTable;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.ShardRoutingState;
+import org.opensearch.cluster.routing.allocation.AllocationService;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.index.Index;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.index.IndexModule;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.opensearch.cluster.metadata.IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING;
+import static org.opensearch.cluster.metadata.IndexMetadata.TIERING_CUSTOM_KEY;
+import static org.opensearch.index.IndexModule.INDEX_STORE_LOCALITY_SETTING;
+import static org.opensearch.index.IndexModule.INDEX_TIERING_STATE;
+
+/**
+ * Service responsible for tiering indices from hot to warm
+ * @opensearch.experimental
+ */
+public class HotToWarmTieringService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(HotToWarmTieringService.class);
+    protected final ClusterService clusterService;
+    protected final IndexNameExpressionResolver indexNameExpressionResolver;
+    protected final AllocationService allocationService;
+    // map to store all the in progress tiering requests
+    private final Map<String, TieringRequestContext> requestUuidToRequestContext = new ConcurrentHashMap<>();
+    // map to store index info of all the indices under-going hot to warm tiering
+    private final Map<Index, IndexTieringInfo> indexTieringInfoMap = new ConcurrentHashMap<>();
```

Comment:
Instead of these 2 we can have a map of `TieringRequestContext` and `IndicesTieringState`. `IndicesTieringState` is essentially a set of in-progress and completed indices. Initially the in-progress set contains all the indices belonging to a `TieringRequest` and completed is empty. As indices completes tiering, it is moved from in-progress to completed set.

Reply:
updated the logic to have just one set of `TieringRequestContext`